packages/e2e-tests/specs/editor/various/change-detection.test.js

```diff
@@ -348,4 +326,22 @@ describe( 'Change detection', () => {
 		// Verify that the post is not dirty.
 		await assertIsDirty( false );
 	} );
+
+	it( 'should not prompt to confirm unsaved changes when trashing an existing post', async () => {
+		// Enter title.
+		await page.type( '.editor-post-title__input', 'Hello World' );
+
+		// Save
+		await saveDraft();
+
+		// Trash post.
+		await openDocumentSettingsSidebar();
+		await page.click( '.editor-post-trash.components-button' );
+
+		// Wait for "Saved" to confirm save complete.
+		await page.waitForSelector( '.editor-post-saved-state.is-saved' );
+
+		// Check that the dialog didn't show.
+		await assertIsDirty( false );
```

Comment:
From your comment at https://github.com/WordPress/gutenberg/pull/18275#discussion_r342644515 :

> This works because refreshing a page with a dialog doesn't get rid of the dialog. So if navigation didn't happen because the dialog stopped it, `assertIsDirty` wouldn't assert `false`.

But the `'dialog'` event handler isn't added until `assertIsDirty` is called, so if the dialog was already present by the time we reach this line, the event callback would never be triggered.

Reply:
True! I missed that.